File: sdk/python/symbolchain/core/nis1/NetworkTimestamp.py
import datetime

from ..NetworkTimestamp import NetworkTimestamp as BasicNetworkTimestamp
from ..NetworkTimestamp import NetworkTimestampDatetimeConverter

EPOCH_TIME = datetime.datetime(2015, 3, 29, 0, 6, 25, tzinfo=datetime.timezone.utc)


class NetworkTimestamp(BasicNetworkTimestamp):
    """Represents a nis network timestamp with second resolution."""

    def add_seconds(self, count):
        return NetworkTimestamp(self.timestamp + count)

    def to_datetime(self, epoch_time=EPOCH_TIME):
        """Converts this timestamp to a datetime."""
        return NetworkTimestampDatetimeConverter(epoch_time, 'seconds').to_datetime(self.timestamp)

    @staticmethod
    def from_datetime(reference_datetime, epoch_time=EPOCH_TIME):
        """Creates a network timestamp from a datetime."""
        return NetworkTimestamp(NetworkTimestampDatetimeConverter(epoch_time, 'seconds').to_difference(reference_datetime))

File: sdk/javascript/src/nem/NetworkTimestamp.js
const BasicNetworkTimestamp = require('../NetworkTimestamp').NetworkTimestamp;
const { NetworkTimestampDatetimeConverter } = require('../NetworkTimestamp');

const createCoverter = epochTime => {
	const EPOCH_TIME = new Date(Date.UTC(2015, 2, 29, 0, 6, 25));
	return new NetworkTimestampDatetimeConverter(epochTime || EPOCH_TIME, 'seconds');
};

/**
 * Represents a NEM network timestamp with millisecond resolution.
 */
class NetworkTimestamp extends BasicNetworkTimestamp {
	/**
	 * Adds a specified number of seconds to this timestamp.
	 * @override
	 * @param {number} count Number of seconds to add.
	 * @returns {NetworkTimestamp} New timestamp that is the specified number of seconds past this timestamp.
	 */
	addSeconds(count) {
		return new NetworkTimestamp(this.timestamp + BigInt(count));
	}

	/**
	 * Converts this timestamp to a datetime.
	 * @param {Date} epochTime Custom network start time (undefined to use mainnet start time).
	 * @returns {Date} Date representation of this timestamp.
	 */
	toDatetime(epochTime = undefined) {
		return createCoverter(epochTime).toDatetime(this.timestamp);
	}

	/**
	 * Creates a network timestamp from a datetime.
	 * @param {Date} referenceDatetime Reference date.
	 * @param {Date} epochTime Custom network start time (undefined to use mainnet start time).
	 * @returns {NetworkTimestamp} Network timestamp representing the reference date.
	 */
	static fromDatetime(referenceDatetime, epochTime = undefined) {
		return new NetworkTimestamp(createCoverter(epochTime).toDifference(referenceDatetime));
	}
}

module.exports = { NetworkTimestamp };
